pandas/core/arrays/integer.py

```diff
@@ -0,0 +1,489 @@
+_unsigned_formatter = lambda x: "{}{}".format(x[0].upper(), x[1:].capitalize())
+
+
+class IntegerDtype(ExtensionDtype):
+    type = None
+    na_value = np.nan
+    kind = 'i'
+    is_integer = True
+    is_signed_integer = True
+    is_unsigned_integer = False
+
+    @classmethod
+    def construct_from_string(cls, string):
+        if string == cls.name:
+            return cls()
+        else:
+            raise TypeError("Cannot construct a '{}' from "
+                            "'{}'".format(cls, string))
+
+    @classmethod
+    def construct_from_string_strict(cls, string):
+        """
+        Strict construction from a string, raise a TypeError if not
+        possible
+        """
+        if string[0] == 'I':
+            return cls.construct_from_string(string)
+        raise TypeError("could not construct PeriodDtype")
+
+
+class UnsignedIntegerDtype(IntegerDtype):
+    kind = 'u'
+    is_signed_integer = False
+    is_unsigned_integer = True
+
+    @classmethod
+    def construct_from_string_strict(cls, string):
+        """
+        Strict construction from a string, raise a TypeError if not
+        possible
+        """
+        if string[0] == 'U':
+            return cls.construct_from_string(string)
+        raise TypeError("could not construct PeriodDtype")
+
+
+def to_integer_array(values):
+    """
+    Parameters
+    ----------
+    values : 1D list-like
+
+    Returns
+    -------
+    infer and return an integer array
+
+    Raises
+    ------
+    TypeError if incompatible types
+    """
+    values = np.array(values, copy=False)
+    kind = 'UInt' if values.dtype.kind == 'u' else 'Int'
+    array_type = "{}{}Array".format(kind, values.dtype.itemsize * 8)
+    try:
+        array_type = getattr(module, array_type)
+    except AttributeError:
+        raise TypeError("Incompatible dtype for {}".format(values.dtype))
+    return array_type(values, copy=False)
+
+
+def coerce_to_array(values, dtype, mask=None, copy=False):
+    """
+    Coerce the input values array to numpy arrays with a mask
+
+    Parameters
+    ----------
+    values : 1D list-like
+    dtype : integer dtype
+    mask : boolean 1D array, optional
+    copy : boolean, default False
+        if True, copy the input
+
+    Returns
+    -------
+    tuple of (values, mask)
+    """
+    values = np.array(values, copy=copy)
+    if is_object_dtype(values):
+        inferred_type = infer_dtype(values)
+        if inferred_type not in ['floating', 'integer',
+                                 'mixed-integer', 'mixed-integer-float']:
+            raise TypeError("{} cannot be converted to an IntegerDtype".format(
+                values.dtype))
+
+    elif not (is_integer_dtype(values) or is_float_dtype(values)):
+        raise TypeError("{} cannot be converted to an IntegerDtype".format(
+            values.dtype))
+
+    if mask is None:
+        mask = isna(values)
+    else:
+        assert len(mask) == len(values)
```

Comment:
should this raise a ValueError instead of AssertionError?

Reply:
no, this is an internal construction error, need to satisfy the input guarantees